synapse/handlers/devicemessage.py

```diff
@@ -167,15 +175,25 @@ async def _check_for_unknown_devices(
 
     async def send_device_message(
         self,
-        sender_user_id: str,
+        requester: Requester,
         message_type: str,
         messages: Dict[str, Dict[str, JsonDict]],
     ) -> None:
+        sender_user_id = requester.user.to_string()
+
         set_tag("number_of_messages", len(messages))
         set_tag("sender", sender_user_id)
         local_messages = {}
         remote_messages = {}  # type: Dict[str, Dict[str, Dict[str, JsonDict]]]
         for user_id, by_device in messages.items():
+            # Ratelimit local cross-user key requests by the sending device.
+            if (
+                message_type == EduTypes.RoomKeyRequest
+                and user_id != sender_user_id
+                and self._ratelimiter.can_do_action(requester.device_id)
```

Comment:
I'm not sure if we consider device IDs to be unique?

```suggestion
                and self._ratelimiter.can_do_action((sender_user_id, requester.device_id))
```

Reply:
Correct, they only need to be unique by user.